# homework 8 
# zhenyang lu

Looks like this is the last three homeworks of it. maybe if this is the last you score the homework for me. Just want to say thanks for the problems session's notes and reviewing the weekly homework.   

# Question 1 

Since the population size is larger than 30, and variance is given by the question, we could use student t distribution. Since the expected sample mean should be the population mean, we could solve this problem by building a 95% CI for the sample mean and see if the population mean is within the range of sample mean's CI.  

After computing the CI for sample mean, population mean is within the CI of sample mean, so we can not reject that, on 95% significance level, this selection constituted a random sample. 

Code used to compute the CI is given below

In [6]:
sample_size = 25
degree_freedom = sample_size -1
population_std = 4
population_mean = 32
sample_mean = 30.4
sample_std = population_std/sqrt(sample_size)
level_of_significance=0.05


print("upper bound:")
sample_mean+qt(1-level_of_significance/2,df=degree_freedom)*sample_std

print("lower bound:")
sample_mean+qt(level_of_significance/2,df=degree_freedom)*sample_std

[1] "upper bound:"


[1] 32.05112

[1] "lower bound:"


[1] 28.74888

[1] 0.8

# Question 2

Using student t distribution with the following parameters,
we could compute p-values as the following code:


In [30]:
sample_mean_1 = 52.5
sample_mean_2 = 55
sample_mean_3 = 57.5

population_mean = 50
population_std = 20
sample_size=64
degree_freedom = sample_size-1

print("p-value 1")
(1-pt(((sample_mean_1-population_mean)/(population_std/sqrt(sample_size))),df=degree_freedom))*2
print("p-value 2")
(1-pt(((sample_mean_2-population_mean)/(population_std/sqrt(sample_size))),df=degree_freedom))*2
print("p-value 3")
(1-pt(((sample_mean_3-population_mean)/(population_std/sqrt(sample_size))),df=degree_freedom))*2

[1] "p-value 1"


[1] 0.321136

[1] "p-value 2"


[1] 0.04981576

[1] "p-value 3"


[1] 0.003863838

# Question 3

Let $\mu$ be the population mean, we are doing the following hypothesis test based on t distribution:

$H_0: \mu = 20 $

$H_a: \mu \ne 20 $

After computing the p value (code following), it's obvious that the p-value is larger than the confidence level, 5%, so we can not reject the null hypothesis that the the population mean is 20 mg on 95% level of confidence. 

In [29]:
print("p-value:") 
pt((19.7-20)/1.3*sqrt(25),df=24)*2

[1] "p-value:"


[1] 0.2599247

# Question 4

# a

let $\mu_{key}$ be the mean of authentication time of users that uses key

let $\mu_{psw}$ be the mean of authentication time of users that uses password

let $\sigma$ be the two population's equal variance, which is the whole population's variance. 

we are trying to test if $\mu_{key}$ - $\mu_{psw}$ is different from 0. So

$H_0: \mu_{key} - \mu_{psw} = 0 $

$H_a: \mu_{key} - \mu_{psw} \ne 0 $


# b 

let $X_{key}$ be  mean authentication time of the users performing "key" authentication

let $X_{psw}$ be  mean authentication time of the users performing "password" authentication

$X_{key}$ ~ $N(\mu_{key},\sigma^2$)

$X_{psw}$ ~ $N(\mu_{psw},\sigma^2$)

$X_{psw} - X_{key}$ ~ $N(\mu_{psw}-\mu_{key},\sigma^2 + \sigma^2$)

$\overline X_{psw} -\overline  X_{key}$ ~ $N(\mu_{psw}-\mu_{key},\frac{\sigma^2}{N_{psd}} + \frac{\sigma^2}{N_{key}}$)

NOTICE: I will use normal distribution to calculate p-value. 

($\overline X_{psw} -\overline  X_{key}$)'s p-value =  $1 - pnorm(\frac{(\overline X_{psw} -\overline  X_{key} - 0}{\sqrt{\frac{\sigma^2}{N_{psw}} + \frac{\sigma^2}{N_{key}}}})$


# c

After computing the p-value, which is much smaller than 1%, we can reject the null hypothesis. We can not deny that there is a difference between users with key authentication and password authentication. Your boss should proceed. 

Code used: 

In [28]:
haha=read.csv("C:\\Users\\zheny\\Dropbox\\courses\\foundation of data analysis (53110)\\lecture8\\logins.csv")
population_var = var(haha$min)
population_mean_key = mean(haha[haha$mechanism == "key",]$min)
population_mean_psw = mean(haha[haha$mechanism == "password",]$min)
population_size_key = length(haha[haha$mechanism == "key",]$min)
population_size_psw = length(haha[haha$mechanism == "password",]$min)

print("p-value: ")
pnorm((population_mean_key-population_mean_psw-0)/(sqrt(population_var/population_size_psw + population_var/population_size_key)))*2

[1] "p-value: "


[1] 3.007334e-139

# question 5

# a



NOTICE: IN THIS PROBLEM, I''M STILL ASSUMING that the mean of the authentication times is approximately normally distributed and that both populations have equal variances.   


let $X_{fa}$ be mean authentication time of faculty

let $X_{st}$ be mean authentication time of staff

let $sigma$ be std deviation of mean authentication time of faculty and staff


$ H_0 :X_{fa} - X_{st} = 0 $

$ H_0 :X_{fa} - X_{st} \ne 0 $


# b

So using normal distribution, we have 

$Z = \frac{\overline X_{st} - \overline X_{fa} - 0} {\sqrt {\frac{\sigma^2}{N_{fa}} +\frac{\sigma^2}{N_{st}}} }$


# c

After computing the p-value, which is much smaller than 1%, we can reject the null hypothesis. We can not deny that there is a mean authentication time difference between faculty and staff. 
 
Code used to calculate the p-value: 

In [32]:
population_var = var(haha$min)
population_mean_faculty = mean(haha[haha$affiliation == "faculty",]$min)
population_mean_staff = mean(haha[haha$affiliation == "staff",]$min)
population_size_faculty = length(haha[haha$affiliation == "faculty",]$min)
population_size_staff = length(haha[haha$affiliation == "staff",]$min)

print("p-value: ")
pnorm((population_mean_faculty-population_mean_staff-0)/(sqrt(population_var/population_size_faculty + population_var/population_size_staff)))*2

[1] "p-value: "


[1] 9.022601e-91

# Question 6

# a 

lik(λ) = $ λe^{−λx_1}...λe^{−λx_n} = λ^ne^{−λ*x_1*x_2...x_n} $, taking the derivative of which, we get MLE $\hat \lambda = \frac{1}{\overline X}$

So $\lambda = $, which can be computed by the following code

In [33]:
data(treering) 
D = diff(c(0, which(treering > 1.6))) 
n = length(D)
1/mean(D)

[1] 0.009962479

# b

the following codes are used to do ks test, as it cen be seen, the p-value is fairly smaller

In [1]:
# b KS test: 

data(treering) 
D = diff(c(0, which(treering > 1.6))) 
n = length(D)

ks.test(rexp(n,1/mean(D)),D) 


Warning message in ks.test(rexp(n, 1/mean(D)), D):
"cannot compute exact p-value with ties"


	Two-sample Kolmogorov-Smirnov test

data:  rexp(n, 1/mean(D)) and D
D = 0.37662, p-value = 3.611e-05
alternative hypothesis: two-sided


[1] 77

# c

function below: 

In [35]:
bootstrap=function(n,v){
  sample(v,n,replace = TRUE)
}

D[D == bootstrap(1,D)]
D[D == bootstrap(1,D)]
D[D == bootstrap(1,D)]

[1] 9

[1] 16 16 16

[1] 5

[1] 1294

# d

Codes used: 

In [2]:
# autocorrelation function 
# lag is the number of lag period
# v is the vector on which 
#   the autocorrelation coefficient is performed
autocorrelation=function(lag,v){
  cor(v[1:(n-lag)],v[(1+lag):(n)])
}

# autocorrelation coefficient of D 
c=autocorrelation(1,D)
print("lag 1 correlation of D")
c

# test random autocorrelation coefficient numbers
n=50
C=autocorrelation(1,bootstrap(n,D))
C=autocorrelation(1,bootstrap(n,D))
print("lag 1 correlation of random sample of D")
C



[1] "lag 1 correlation of D"


[1] 0.08491135

ERROR: Error in is.data.frame(y): 没有"bootstrap"这个函数


In [58]:
# part d continue

# result contains the distribution of C 
result =vector("numeric")

# loop specify the number of C, which is the number of times
# a random sample is drawn and autocorrelation coefficient 
# number is computed
 
loop = 10000

for (i in c(1:loop)){
   result = append(result,autocorrelation(1,bootstrap(n,D)))
}

print("p-value: ")
2* min(length(result[result>c])/length(result),length(result[result<c])/length(result))


[1] "p-value: "


[1] 0.2668

After simulating the above code several times, I got constantly p-value between 0.24 - 0.27. So the conclusion is we can not reject the null hypothesis that the values of D are drawn independently. 